constructor/signing.py

```diff
@@ -0,0 +1,220 @@
+            "sha256"
+        )
+        file_digest = os.environ.get(
+            "AZURE_SIGNTOOL_FILE_DIGEST",
+            "sha256"
+        )
+
+        command = (
+            f"{self.executable} sign -v"
+            ' -kvu %AZURE_SIGNTOOL_KEY_VAULT_URL%'
+            ' -kvc %AZURE_SIGNTOOL_KEY_VAULT_CERTIFICATE%'
+            f" -tr {timestamp_server}"
+            f" -td {timestamp_digest}"
+            f" -fd {file_digest}"
+        )
+        # There are three ways to sign:
+        #   1. Access token
+        #   2. Secret (requires tenant ID)
+        #   3. Managed identity (requires prior login to Azure)
+        if "AZURE_SIGNTOOL_KEY_VAULT_ACCESSTOKEN" in os.environ:
+            logger.info("AzureSignTool: signing binary using access token.")
+            command += ' -kva "%AZURE_SIGNTOOL_KEY_VAULT_ACCESSTOKEN%"'
+        elif "AZURE_SIGNTOOL_KEY_VAULT_SECRET" in os.environ:
+            # Authentication via secret required client and tenant ID
+            logger.info("AzureSignTool: signing binary using secret.")
+            required_env_vars = (
+                "AZURE_SIGNTOOL_KEY_VAULT_CLIENT_ID",
+                "AZURE_SIGNTOOL_KEY_VAULT_TENANT_ID",
+            )
+            check_required_env_vars(required_env_vars)
+            command += (
+                " -kvi %AZURE_SIGNTOOL_KEY_VAULT_CLIENT_ID%"
+                " -kvt %AZURE_SIGNTOOL_KEY_VAULT_TENANT_ID%"
+                " -kvs %AZURE_SIGNTOOL_KEY_VAULT_SECRET%"
+            )
+        else:
+            # No token or secret found, assume managed identity
+            logger.info("AzureSignTool: signing binary using managed identity.")
+            command += " -kvm"
+        return command
+
+    def verify_signing_tool(self):
+        self._verify_tool_is_available()
+        check_call([self.executable, "--help"], stdout=PIPE, stderr=PIPE)
+
+    def verify_signature(self, installer_file: Union[str, Path]):
+        """Use Powershell to verify signature.
+
+        For available statuses, see the Microsoft documentation:
+        https://learn.microsoft.com/en-us/dotnet/api/system.management.automation.signaturestatus
+        """
+        if shutil.which("powershell") is None:
+            logger.error("Could not verify signature: PowerShell not found.")
+            return
+        command = (
+            f"$sig = Get-AuthenticodeSignature -LiteralPath {installer_file};"
+            "$sig.Status.value__;"
+            "$sig.StatusMessage"
+        )
+        proc = run([
+                "powershell",
+                "-c",
+                command,
+            ],
+            capture_output=True,
+            text=True,
+        )
+        # The return code will always be 0,
+        # but stderr will be non-empty on errors
+        if proc.stderr:
+            raise RuntimeError(f"Signature verification failed.\n{proc.stderr}")
+        try:
+            status, status_message = proc.stdout.strip().split("\n")
+            status = int(status)
```

Comment:
Can this fail if status is empty or None?

Reply:
I'm not sure how this could be `None` unless `proc.stdout` can somehow be `None`.

If `status` is and empty string, `int(status)` raises a `ValueError`, which is caught further down.